lib/networkx/digraph.rb

```diff
@@ -0,0 +1,218 @@
+    #   graph.remove_node("Noida")
+    #
+    # @param node [Object] the node to be removed
+    def remove_node(node)
+      raise KeyError, "Error in deleting node #{node} from Graph." unless @nodes.key?(node)
+
+      neighbours = @adj[node]
+      neighbours.each_key { |k| @pred[k].delete(node) }
+      @pred[node].each_key do |k|
+        @adj[k].delete(node)
+      end
+
+      @pred.delete(node)
+      @adj.delete(node)
+      @nodes.delete(node)
+    end
+
+    # Removes edge from the graph
+    #
+    # @example
+    #   graph.remove_edge('Noida', 'Bangalore')
+    #
+    # @param node_1 [Object] the first node of the edge
+    # @param node_2 [Object] the second node of the edge
+    def remove_edge(node_1, node_2)
+      raise KeyError, "#{node_1} is not a valid node." unless @nodes.key?(node_1)
+      raise KeyError, "#{node_2} is not a valid node" unless @nodes.key?(node_2)
+      raise KeyError, 'The given edge is not a valid one.' unless @adj[node_1].key?(node_2)
+
+      @adj[node_1].delete(node_2)
+      @pred[node_2].delete(node_1)
+    end
+
+    # Clears the graph
+    #
+    # @example
+    #   graph.clear
+    def clear
+      super
+
+      @pred.clear
+    end
+
+    # Returns number of edges
+    #
+    # @example
+    #   graph.number_of_edges
+    def number_of_edges
+      num = 0
+      @adj.each { |_, v| num += v.length }
+      num
+    end
+
+    # Returns number of edges if is_weighted is false
+    # or returns total weight of all edges
+    #
+    # @example
+    #   graph.size(true)
+    #
+    # @praram is_weighted [Bool] if we want weighted size of unweighted size
+    def size(is_weighted=false)
+      if is_weighted
+        graph_size = 0
+        @adj.each do |_, hash_val|
+          hash_val.each { |_, v| graph_size += v[:weight] if v.key?(:weight) }
+        end
+        return graph_size
+      end
+      number_of_edges
+    end
+
+    # Returns in-degree of a given node
+    #
+    # @example
+    #   graph.in_degree(node)
+    #
+    # @param node [Object] the node whose in degree is to be calculated
+    def in_degree(node)
+      @pred[node].length
+    end
+
+    # Returns out-degree of a given node
+    #
+    # @example
+    #   graph.out_degree(node)
+    #
+    # @param node [Object] the node whose out degree is to be calculated
+    def out_degree(node)
+      @adj[node].length
+    end
+
+    # Returns the reversed version of a the graph
+    #
+    # @example
+    #   graph.reverse
+    def reverse
+      new_graph = NetworkX::DiGraph.new(@graph)
+      @nodes.each_key { |k| new_graph.add_node(k, @nodes[k]) }
+      @adj.each_key do |k|
+        @adj[k].each_key { |v| new_graph.add_edge(v, k, @adj[k][v]) }
+      end
+      new_graph
+    end
+
+    # Returns the undirected version of the graph
+    #
+    # @example
+    #   graph.to_undirected
+    def to_undirected
+      new_graph = NetworkX::Graph.new(@graph)
+      @nodes.each_key { |k| new_graph.add_node(k, @nodes[k]) }
```

Comment:
Any specific reason for not using the more obvious `each` rather than `each_key` when we anyway have to access the value (`@nodes[k]`) too? Maybe change it to:

```ruby
@nodes.each { |k, v| new_graph.add_node(k, v) }
```